pd/src/consensus/worker.rs

```diff
@@ -247,164 +241,235 @@ impl Worker {
             .height
             .try_into()
             .expect("height should be nonnegative");
-        let epoch = pending_block.set_height(height);
+        let epoch = pending_block.set_height(
+            height,
+            self.state
+                .private_reader()
+                .chain_params_rx()
+                .borrow()
+                .epoch_duration,
+        );
 
         tracing::debug!(?height, ?epoch, end_height = ?epoch.end_height());
 
+        // Find out which validators were slashed in this block
+        let slashed_validators = pending_block
+            .validator_state_changes
+            .iter()
+            .filter(|p| matches!(p.1, ValidatorState::Slashed))
+            .map(|p| p.0) // get the validator ID
+            .collect::<Vec<_>>();
+
+        // Immediately revert notes and nullifiers immediately from slashed validators in this block
```

Comment:
We immediately revert quarantined notes and nullifiers undelegated from a slashed validator; we do not wait for the next epoch boundary.